### PR TITLE
feat!(rust,python): Create engine from dataframe

### PR DESCRIPTION
**Pull request by schmidmt**

### CODE REVIEW

File: pylace/README.md [hunk 1]
```diff
@@ -31,9 +31,11 @@ file. The following lace functions are supported:
 
 ```python
 import lace
+import polars as pl
 
 # The required files can be found here: https://github.com/promised-ai/lace/tree/master/pylace/lace/resources/datasets/satellites
-engine = lace.Engine(data_source="data.csv", codebook="codebook.yaml")
+df = pl.read_csv("./data.csv")
+engine = lace.Engine.load((df, codebook_builder="codebook.yaml")
```
Comment (BaxterEaves) [inline pylace/README.md:38]: I think the arg name should stay `codebook`.

File: pylace/lace/engine.py [hunk 2]
```diff
@@ -39,66 +39,113 @@ def __init__(self, df: pl.DataFrame, linkage: "np.ndarray", figure=None):
...
+        """
+        Create a new ``Engine`` with its internal representation.
+
+        In general, you should use ``Engine.from_dataframe`` or ``Engine.load``
+        instead.
+        """
+        self.engine = core_engine
+
+    @classmethod
+    def from_dataframe(
```
Comment (BaxterEaves) [inline pylace/lace/engine.py:54]: `from_df`? Happy to take votes on this.

File: pylace/lace/engine.py [hunk 3]
```diff
@@ -39,66 +39,113 @@ def __init__(self, df: pl.DataFrame, linkage: "np.ndarray", figure=None):
...
     """The cross-categorization model with states and data."""
 
-    def __init__(self, *args, **kwargs):
+    engine: core.CoreEngine
+
+    def __init__(self, core_engine: core.CoreEngine) -> None:
+        """
+        Create a new ``Engine`` with its internal representation.
+
+        In general, you should use ``Engine.from_dataframe`` or ``Engine.load``
```
Comment (BaxterEaves) [inline pylace/lace/engine.py:48]: you *should* use -> you *will* use

File: pylace/lace/engine.py [hunk 4]
```diff
@@ -39,66 +39,113 @@ def __init__(self, df: pl.DataFrame, linkage: "np.ndarray", figure=None):
...
+        instead.
+        """
+        self.engine = core_engine
+
+    @classmethod
+    def from_dataframe(
+        cls,
+        dataframe: Union[pd.DataFrame, pl.DataFrame],
+        codebook_builder: Optional[
```
Comment (BaxterEaves) [inline pylace/lace/engine.py:57]: Let's keep the arg name as `codebook`.

File: pylace/lace/engine.py [hunk 5]
```diff
@@ -39,66 +39,113 @@ def __init__(self, df: pl.DataFrame, linkage: "np.ndarray", figure=None):
...
-        >>> engine = Engine(metadata="metadata.lace")  # doctest: +SKIP
+        >>> from lace import Engine
+        >>> import polars as pl
+        >>> df = pl.DataFrame({
+        >>>    "list_a": [[1, 2, 3], [5, 5]],
+        >>>    "list_b": [[1, 2, 3, 8], [5, 1, 1]]
+        >>> })
+        >>> engine = Engine.from_dataframe(df)
+
+        Create a new ``Engine`` with specific codebook inference rules
```
Comment (BaxterEaves) [inline pylace/lace/engine.py:99]: Thanks for including this in the docstring!

File: pylace/lace/engine.py [hunk 6]
```diff
@@ -39,66 +39,113 @@ def __init__(self, df: pl.DataFrame, linkage: "np.ndarray", figure=None):
...
+        >>> })
+        >>> engine = Engine.from_dataframe(df, CodebookBuilder.infer(
+        >>>     cat_cutoff=2,
+        >>> ))
 
-        Create a new Engine with default codebook. The start state is drawn from
-        the probabilistic cross-categorization prior.
+        """
+        if isinstance(dataframe, pd.DataFrame):
+            dataframe = pl.DataFrame._from_pandas(dataframe)
```
Comment (BaxterEaves) [inline pylace/lace/engine.py:112]: Does this function convert the pandas index into a column in the polars series or is it removed?

File: pylace/lace/engine.py [hunk 7]
```diff
@@ -1025,46 +1072,45 @@ def surprisal(
...
-            Proposed values for each cell. Must have an entry for each entry in
-            `rows`. If `None`, the existing values are used.
+            Proposed values for each cell. Must have an entry for each entry in `rows`. If `None`, the existing values are used.
         state_ixs: List[int], optional
             An optional list specifying which states should be used in the
             surprisal computation. If `None` (default), use all states.
 
         Returns
-        -------
+        - ------
```
Comment (BaxterEaves) [inline pylace/lace/engine.py:1091]: What is going on with these spaces everywhere?

File: pylace/lace/engine.py [hunk 8]
```diff
@@ -1376,17 +1422,17 @@ def impute(
...
-            .. math::
+            .. math: :
               JS(X_1, X_2, ..., X_S)
 
           * unc_type='pairwise_kl' computes the mean of the Kullback-Leibler
             divergences between pairs of state imputation distributions.
 
-            .. math::
-              \frac{1}{S^2 - S} \sum_{i=1}^S \sum{j \in \{1,..,S\} \setminus i} KL(X_i | X_j)
+            .. math: :
```
Comment (BaxterEaves) [inline pylace/lace/engine.py:1431]: What is going on? Looks like your linter is acting up.

File: pylace/src/lib.rs [hunk 9]
```diff
@@ -6,113 +6,141 @@ use std::collections::HashMap;
...
+            } => df_to_codebook(df, cat_cutoff, alpha_prior_opt, no_hypers)
+                .map_err(|e| {
+                    PyValueError::new_err(format!(
+                        "Failed to infer the Codebook. Error: {e}"
+                    ))
+                }),
+        }
+    }
+
+    fn __repr__(&self) -> String {
```
Comment (BaxterEaves) [inline pylace/src/lib.rs:116]: Thanks for implementing this!

**Review by BaxterEaves — CHANGES_REQUESTED** [review]
Looks mostly good -- found an error, have some naming considerations.

Your linter or something has gone nuts with the spacing in `engine.py`. It has added and removed a lot of spaces in doc string because, I think it lost track of what is code and what is comments?

File: lace/src/data/data_source.rs [hunk 10]
```diff
@@ -17,10 +18,46 @@ pub enum DataSource {
...
     Parquet(PathBuf),
+    /// Polars DataFrame
+    Polars(DataFrame),
     /// Empty (A void datasource).
     Empty,
 }
 
+/// Error when extension is not CSV, JSON, IPC, or Parquet
+#[derive(Clone, Debug)]
+pub struct UnknownExtension(pub Option<OsString>);
```
[the remark this answers is not in the document]
Comment (schmidmt) [reply lace/src/data/data_source.rs:29]: Done!

File: pylace/README.md (hunk 1 above)
Comment (schmidmt) [reply pylace/README.md]: Agreed. Fixed.

File: pylace/lace/engine.py (hunk 6 above)
Comment (schmidmt) [reply pylace/lace/engine.py]: Fixed.

**Review by BaxterEaves — CHANGES_REQUESTED** [review]
Please bump version in lace_stats, lace, and pylace and include changes in the changelogs.

**Pull request merged**